pkg/kubelet/kuberuntime/fake_kuberuntime.go

```diff
@@ -0,0 +1,482 @@
+	return true
+}
+
+func (r *fakeKubeRuntime) ListPodSandbox(filter *runtimeApi.PodSandboxFilter) ([]*runtimeApi.PodSandboxListItem, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.called = append(r.called, "ListPodSandbox")
+
+	result := make([]*runtimeApi.PodSandboxListItem, 0)
+	for id, s := range r.sandboxes {
+		if filter != nil {
+			if filter.Id != nil && filter.GetId() != id {
+				continue
+			}
+			if filter.Name != nil && filter.GetName() != s.config.GetName() {
+				continue
+			}
+			if filter.State != nil && filter.GetState() != s.state {
+				continue
+			}
+			if filter.LabelSelector != nil && !filterInLabels(filter.LabelSelector, s.config.Labels) {
+				continue
+			}
+		}
+
+		result = append(result, &runtimeApi.PodSandboxListItem{
+			Id:        &id,
+			Name:      s.config.Name,
+			State:     &s.state,
+			CreatedAt: &s.createdAt,
+			Labels:    s.config.Labels,
+		})
+	}
+
+	return result, nil
+}
+
+func (r *fakeKubeRuntime) CreateContainer(podSandBoxID string, config *runtimeApi.ContainerConfig, sandboxConfig *runtimeApi.PodSandboxConfig) (string, error) {
+	r.Lock()
+	defer r.Unlock()
+
+	r.called = append(r.called, "CreateContainer")
+
+	imageID := ""
+	for _, img := range r.images {
+		if stringInSlice(config.Image.GetImage(), img.RepoTags) {
+			imageID = img.GetId()
+			break
+		}
+	}
+	if imageID == "" {
+		return "", fmt.Errorf("Image %s not found", config.Image.GetImage())
+	}
+
+	labelledInfo := getContainerInfoFromLabel(config.Labels)
+	containerID := getFakeContainerID(labelledInfo.Name, labelledInfo.PodName, labelledInfo.PodNamespace)
+	r.containers[containerID] = &containerWithState{
+		createdAt:       time.Now().Unix(),
+		containerID:     containerID,
+		podSandBoxID:    podSandBoxID,
+		containerConfig: config,
+		sandboxConfig:   sandboxConfig,
+		imageID:         imageID,
+		state:           runtimeApi.ContainerState_CREATED,
+	}
+
+	return containerID, nil
+}
+
+func (r *fakeKubeRuntime) StartContainer(rawContainerID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.called = append(r.called, "StartContainer")
+
+	if c, ok := r.containers[rawContainerID]; ok {
+		c.state = runtimeApi.ContainerState_RUNNING
+	} else {
+		return fmt.Errorf("Container not found")
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) StopContainer(rawContainerID string, timeout int64) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.called = append(r.called, "StopContainer")
+
+	if c, ok := r.containers[rawContainerID]; ok {
+		c.state = runtimeApi.ContainerState_EXITED
+	} else {
+		return fmt.Errorf("Container not found")
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) RemoveContainer(rawContainerID string) error {
+	r.Lock()
+	defer r.Unlock()
+
+	r.called = append(r.called, "RemoveContainer")
+
+	if _, ok := r.containers[rawContainerID]; ok {
+		delete(r.containers, rawContainerID)
+	}
+
+	return nil
+}
+
+func (r *fakeKubeRuntime) ListContainers(filter *runtimeApi.ContainerFilter) ([]*runtimeApi.Container, error) {
+	r.Lock()
```

Comment:
@resouer `fakeKubeRuntime` is only for unit testing now. I think its ok with `Mutex`'s `Lock`.